src/docfx/lib/ProcessUtility.cs

```diff
@@ -98,5 +107,56 @@ public static Task<string> Execute(string fileName, string commandLineArgs, stri
 
             return tcs.Task;
         }
+
+        /// <summary>
+        /// Provide a process lock function based on locking file
+        /// </summary>
+        /// <param name="action">The action you want to lock</param>
+        /// <param name="lockPath">The lock file path, default is a file with GUID name</param>
+        /// <param name="retry">The retry count, default is 600 times</param>
+        /// <param name="retryTimeSpanInterval">The retry interval, default is 1 seconds</param>
+        /// <returns>The task status</returns>
+        public static async Task ProcessLock(Func<Task> action, string lockPath, int retry = 600, TimeSpan? retryTimeSpanInterval = null)
+        {
+            Debug.Assert(!string.IsNullOrEmpty(lockPath));
+            Debug.Assert(!PathUtility.FilePathHasInvalidChars(lockPath));
+            Directory.CreateDirectory(Path.GetDirectoryName(lockPath));
+
+            using (var lockFile = await AcquireFileStreamLock(lockPath, retry, retryTimeSpanInterval ?? TimeSpan.FromSeconds(1)))
+            {
+                try
+                {
+                    await action();
+                }
+                finally
+                {
+                    File.Delete(lockPath);
+                }
+            }
+        }
+
+        private static async Task<FileStream> AcquireFileStreamLock(string lockPath, int retry, TimeSpan retryTimeSpanInterval)
```

Comment:
>AcquireFileStreamLock [](start = 46, length = 21)

if it is some kind of lock, is it more reasonable to provide a timeout instead?

Reply:
i want to use retryCount + timespanInterval instead of timeout, timeout would be hard to control the frequency of checking and it's not really accurate

---
In reply to: [182647781](https://github.com/dotnet/docfx/pull/2651#discussion_r182647781) [](ancestors = 182647781)